pkg/certgenerator/v1beta1/generator.go

```diff
@@ -0,0 +1,245 @@
+/*
+Copyright 2022 The Kubeflow Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certgenerator
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"errors"
+	"fmt"
+	"math/big"
+	"strings"
+	"time"
+
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	configv1beta1 "github.com/kubeflow/katib/pkg/apis/config/v1beta1"
+	"github.com/kubeflow/katib/pkg/controller.v1beta1/consts"
+)
+
+var (
+	errServiceNotFound      = errors.New("unable to locate controller service")
+	errCertCheckFail        = errors.New("failed to check if certs already exist")
+	errCreateCertFail       = errors.New("failed to create certs")
+	errCreateCertSecretFail = errors.New("failed to create secret embedded certs")
+	errInjectCertError      = errors.New("failed to inject certs into WebhookConfigurations")
+)
+
+// CertGenerator is the manager to generate certs.
+type CertGenerator struct {
+	namespace          string
+	webhookServiceName string
+	webhookSecretName  string
+	kubeClient         client.Client
+	certsReady         chan struct{}
+
+	certs             *certificates
+	fullServiceDomain string
+}
+
+var _ manager.Runnable = &CertGenerator{}
+var _ manager.LeaderElectionRunnable = &CertGenerator{}
+
+func (c *CertGenerator) Start(ctx context.Context) error {
+	if err := c.generate(ctx); err != nil {
+		return err
+	}
+	// Close a certsReady means start to register controllers to the manager.
+	close(c.certsReady)
```

Comment:
Minor. closing certsReady looks confusing to me.  From the name, i assume that certsReady channel will receive a message when certs are ready.  But, we are closing the channel instead. Either we send a message instead of closing or rename the channel?

Reply:
I think we should close the channel since sending data to the channel is some overhead rather than closing the channel.